core/internal/peerstate/peerstate.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2018 NEC Laboratories Europe GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package peerstate provides means to interact with a representation
+// of the state maintained by the replica for each peer replica.
+package peerstate
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/nec-blockchain/minbft/usig"
+)
+
+// Provider returns an instance of state representation associated
+// with a peer replica given its ID. It is safe to invoke
+// concurrently.
+type Provider func(replicaID uint32) State
+
+// NewProvider creates an instance of Provider
+func NewProvider() Provider {
+	var (
+		lock sync.Mutex
+		// Replica ID -> replica state
+		peerStates = make(map[uint32]State)
+	)
+
+	return func(replicaID uint32) State {
+		lock.Lock()
+		defer lock.Unlock()
+
+		state := peerStates[replicaID]
+		if state == nil {
+			state = New()
+			peerStates[replicaID] = state
+		}
+
+		return state
+	}
+}
+
+// State represents the state maintained by the replica for each peer
+// replica. All methods are safe to invoke concurrently.
+//
+// AcceptUI accepts a valid USIG unique identifier ui. A UI cannot be
+// accepted until all previous UIs generated by that USIG instance
+// have been accepted and committed. If the UI cannot be accepted
+// immediately, it will block until the UI can be accepted. The return
+// value new indicates if the UI has not been accepted before.
+//
+// CommitUI records the last accepted USIG unique identifier ui as
+// committed so that the next UI can be accepted. Each UI can be
+// committed one time only.
+type State interface {
+	AcceptUI(ui *usig.UI) (new bool)
+	CommitUI(ui *usig.UI) error
+}
```

Comment:
I'm actually I bit worried about the fact that the terms "accept" and "commit" may be too overloaded. Those also heavily used in MinBFT paper in various contexts like "accept message", "accept request", "COMMIT message". If we also count PBFT paper, that most of our users would be likely familiar with, it comes even worse. PBFT paper uses terms like "locally committed request", "committed request" etc.

So I just thought of possible change in terms to express the two-stage UI handling:

- `AcceptUI` → `CaptureUI`, or `AcquireUI`
- `CommitUI` → `ReleaseUI`, or `UnblockUI`, or `UnlockUI`

What do you think?

Reply:
I feel `CaptureUI` and `ReleaseUI` sounds good (matched well).

Reply:
Me too. I opt for `CaptureUI` and `ReleaseUI`

Reply:
:+1: Thanks, I'll change those names in v2.